Implement Image and Table support + Laravel Pint addition

igorsgm opened the pull request:
Hello @roelmagdaleno. Thank you for this useful package. I hope my PR will be a valuable contribution 🙏
 
## Summary

  - Implemented Notion Image handling with dedicated processor and fallback external URL detection
  - Implemented Notion Table handling
  - Improved CommonMark integration with updated dependencies and better block rendering
  - Added comprehensive code formatting standards with Laravel Pint

## Changes

  - Image Processing: Added MarkdownImageProcessor for better image URL handling and file validation
  - Updated Dependencies: Upgraded CommonMark to v2.7.0, Pest to v3.8.2, and added Laravel Pint v1.22
  - Code Quality: Implemented strict formatting rules with comprehensive Pint configuration
  - Testing: Added extensive image processing tests and improved test coverage
  - UTF-8 Handling: Enhanced invalid character validation to prevent encoding issues


The changes maintain backward compatibility while significantly improving the package's robustness and code quality standards.

igorsgm committed:
Conversion improvements (#1)

* Bump dependencies + installing Laravel Pint

* .gitignore added

* Opinionated pint.json from nunomaduro/essentials added

* Pest test fix

* Add image processing and update annotations handling

igorsgm committed:
Conversion improvements (#2)

* Bump dependencies + installing Laravel Pint

* .gitignore added

* Opinionated pint.json from nunomaduro/essentials added

* Pest test fix

* Add image processing and update annotations handling

igorsgm committed:
Conversion improvements (#3)

* Bump dependencies + installing Laravel Pint

* .gitignore added

* Opinionated pint.json from nunomaduro/essentials added

* Pest test fix

* Add image processing and update annotations handling

* bump version

* Pest tests implemented

igorsgm committed:
Image validation (#4)

* wip

* wip

roelmagdaleno commented:
Hey @igorsgm, I'll review it this weekend. Thank you.

igorsgm committed:
Table support

roelmagdaleno commented on src/Converter/NotionBlocksRenderer.php:
@igorsgm The image processor in this class is out place here.

The `NotionBlocksRenderer` class is like the entry point of the entire process and must call the Notion blocks functionality outside of it. You can see there's no other Notion block call within the class. So, calling the image processor there breaks the pattern.

Hope that makes sense, thanks.

roelmagdaleno commented on src/NotionBlocks/Image.php:
```diff
@@ -0,0 +1,107 @@
+    public function object(): array
+    {
+        $url = $this->node->getUrl();
+        $isExternalUrl = filter_var($url, FILTER_VALIDATE_URL) !== false;
```
The `$isExternalUrl` exist in the `ImageValidator` class, why duplicated here? Also, this line is called at the beginning of this method but being used almost at the end.

roelmagdaleno commented on src/NotionBlocks/Image.php:
@igorsgm The `Image` class is good but I detect at least 3 image types: invalid image, external image (from link), and an image file.

Would be cool to add an Enum class (or maybe separated classes) for those images, ex: `ImageType`, that'd have 3 methods:

- `invalid()`
- `external()`
- `file()`

All those methods will return their own array, so it's easy to edit the data if Notion decides to change, let's say the JSON structure for the external or a file image.

What do you think? Thanks.

roelmagdaleno commented on src/Validation/ImageValidator.php:
@igorsgm The `ImageValidator` class looks good. I'd move the class to a `Validators` folder instead of `Validation`. The `ImageValidator` class has the word `Validator`, so I expect that to live inside a folder with the same name.

This is more for a good developer experience. Let's keep a consistency between class names and folders.

roelmagdaleno commented on src/Validation/ImageValidator.php:
@igorsgm Why have the `ImageValidator` with static methods? I mean, we could have a method inside the `Image` class like:

````php

<?php

private function isValid() {
     return (new ImageValidator())->isValidNotionImage($this->node->getUrl());
}
````

Then, you'd call it like this:

```php
<?php

if (!$this->isValid($url)) {}
```

IMO, static methods are going to be used by several classes, and it's only used by a single class and tests.

About the tests, those are using the static methods and feel hacky because the real use case is to get the images from a markdown file and you're just putting the URLs from an array:

```php
<?php

test('ImageValidator validates supported image extensions', function () {
    $validUrls = [
        'https://example.com/image.jpg',
        'https://example.com/image.jpeg',
        'https://example.com/image.png',
        'https://example.com/image.gif',
        'https://example.com/image.svg',
        'https://example.com/image.bmp',
        'https://example.com/image.tif',
        'https://example.com/image.tiff',
        'https://example.com/image.heic',
    ];

    foreach ($validUrls as $url) {
        expect(ImageValidator::isValidNotionImage($url))->toBeTrue("URL $url should be valid");
    }
});
```

Hope this makes sense. Thanks.

roelmagdaleno commented on src/NotionBlocks/Table.php:
```diff
@@ -0,0 +1,116 @@
+     *
+     * @return bool True if table has headers
+     */
+    private function detectHasHeader(): bool
```
@igorsgm The method name can be named as `hasHeader`. The word `detect` is adding too much cognitive friction.

roelmagdaleno reviewed (changes requested):
@igorsgm Thanks for this PR.

Adding the image and table components are good features.

I've reviewed all files and found some code that could be improved, let me know your thoughts about it. You'll find my comments on the files.

Also, if possible, let's use defensive programming with guard clauses. That means to remove the `else` statements and nested `if` conditions. We're trying to avoid the callback/condition hell and improve code readability.

Here's a [good article](https://refactoring.guru/replace-nested-conditional-with-guard-clauses) that explains it all.

Thank you. Let me know any feedback.

igorsgm committed:
Image validation refactoring and code improvements

- Bump minimum PHP version requirement from 8.0 to 8.1 for enhanced language features (enum)
- Refactor image validation by creating ImageType enum to centralize image block creation logic
- Convert ImageValidator from static class to instance-based with proper encapsulation
- Improve code readability with early returns and guard clauses in MarkdownImageProcessor and NotionBlocksRenderer
- Move ImageValidator from Validation to Validators namespace for better organization
- Optimize control flow structures using continue statements to reduce nesting depth
- Make RichText defaultObject method public static for better reusability across components
- Enhance table processing logic with cleaner conditional handling in Table class
- Update constructor injection patterns to use readonly properties for better performance
- Rename ImageValidationTest to ImageValidatorTest to match new class structure

igorsgm commented:
Hi @roelmagdaleno,

  Thank you for the excellent feedback on this PR! I've successfully addressed all your suggestions:

  ## ✅ **Changes Made:**

  ### 1. **Renamed `detectHasHeader` to `hasHeader`** in Table.php
  - Improved method naming for better readability

  ### 2. **Moved ImageValidator to `Validators/` folder**
  - Better consistency between class names and folder structure
  - Updated all imports accordingly

  ### 3. **Converted ImageValidator from static to instance methods**
  - Added private `$validator` property to `Image` class
  - Created `isValid()` method as suggested
  - Eliminated duplicate `$isExternalUrl` logic

  ### 4. **Created ImageType enum with specialized methods**
  - Added `INVALID`, `EXTERNAL`, `FILE` cases
  - Moved caption logic to enum using `RichText::defaultObject()`

  ### 5. **Applied defensive programming with guard clauses**
  - Eliminated nested if-else statements throughout codebase
  - Applied early returns to reduce nesting
  - Updated: `NotionBlocksRenderer.php`, `MarkdownImageProcessor.php`, `ImageValidator.php`, `Table.php`

  ## 📝 **Regarding the Image Processor Placement:**

  The image processor in `NotionBlocksRenderer` serves a specific architectural purpose. It needs to analyze CommonMark
   nodes during the rendering phase to:
  - Extract images from complex nested node structures
  - Handle image-like links (like `![alt](url)` syntax)
  - Skip paragraphs that contain only images
  - Access internal CommonMark node data during conversion

  Moving it outside would require exposing internal CommonMark node structures and complicate the rendering pipeline.
  The current design maintains clean separation of concerns while allowing necessary access to node data during the
  conversion process.

  ## ✅ **Results:**
  - **All 82 tests pass** - no breaking changes
  - Better design patterns with improved separation of concerns
  - Reduced complexity through guard clauses
  - Enhanced maintainability with consistent `RichText::defaultObject()` usage

  Thank you for the thoughtful review and excellent suggestions!

igorsgm committed:
Update character limit for rich text objects and add tests for code block splitting
